### PR TITLE
Add one time trigger.

### PR DESCRIPTION
**Pull request by rizakii**
- Add ScheduleOnce predicate
- Add ScheduleOnce event source

### CODE REVIEW

**Comment by rizakii** [comment]
@satoshikumano please do the design review.

**Comment by rizakii** [comment]
closed because need to address other issues.

**Comment by rizakii** [comment]
@satoshikumano please review the implementations.

File: thingif/src/main/java/com/kii/thingif/trigger/ScheduleOncePredicate.java [hunk 1]
```diff
@@ -14,7 +14,7 @@
      * @throws IllegalArgumentException when scheduleAt is not in the future.
```
Comment (satoshikumano) [inline thingif/src/main/java/com/kii/thingif/trigger/ScheduleOncePredicate.java]: We will check at execution of postTrigger?

File: thingif/src/main/java/com/kii/thingif/ThingIFAPI.java [hunk 2]
```diff
@@ -589,6 +590,16 @@ public Command getCommand(
         return new Pair<List<Command>, String>(commands, nextPaginationKey);
     }
 
+    private void validatePredicate(Predicate predicate) throws IllegalArgumentException {
```
Comment (satoshikumano) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPI.java]: Let's move it to Predicate.

**Comment by satoshikumano** [comment]
Hmm. Defer validation of predicate does not seems a good idea.

**Comment by satoshikumano** [comment]
Let's not validate the timestamp in client side.
Developer man not frequently set past time for trigger.

**Comment by rizakii** [comment]
@satoshikumano please review Tests.

**Comment by satoshikumano** [comment]
Do not use admin feature in the test.

**Comment by rizakii** [comment]
@satoshikumano review indication is fixed, please confirm.

File: thingif/src/androidTest/java/com/kii/thingif/ThingIFAPI_PatchTriggerTest.java [hunk 3]
```diff
@@ -37,8 +39,7 @@
  */
 @RunWith(AndroidJUnit4.class)
 public class ThingIFAPI_PatchTriggerTest extends ThingIFAPITestBase {
-    @Test
-    public void patchTriggerWithCommandTest() throws Exception {
+    private void _patchTriggerWithCommandTest(Predicate predicate) throws Exception {
```
Comment (satoshikumano) [inline thingif/src/androidTest/java/com/kii/thingif/ThingIFAPI_PatchTriggerTest.java]: In java method name starting '_' is not so popular convention.

Can we choose different name?

File: thingiftest/src/androidTest/java/com/kii/thingiftest/largetests/TriggerTest.java [hunk 4]
```diff
@@ -34,17 +36,22 @@
...
 @RunWith(AndroidJUnit4.class)
+
 public class TriggerTest extends LargeTestCaseBase {
-    @Test
-    public void basicStatePredicateTriggerTest() throws Exception {
+    private interface PredicateChecker <T extends Predicate>{
+        void checkPredicate (T predicate);
+    }
+
+    private void _basicTriggerTest(LinkedList<Predicate> predicates, LinkedList<PredicateChecker> checkers) throws Exception{
```
Comment (satoshikumano) [inline thingiftest/src/androidTest/java/com/kii/thingiftest/largetests/TriggerTest.java]: Lets not use '_'.

**Comment by satoshikumano** [comment]
Updated Test Case Coverage strategy for this feature:
# Small Test:
Post Trigger (existing)
Patch Trigger
Get Trigger
List Trigger

# Large Test
Post and Get (In new Test Case class)

Whether the triggersWhat is command or server code is orthogonal in this case, no need to test both for one time predicate.

**Comment by rizakii** [comment]
@satoshikumano review indications are fixed.
Please confirm

**Pull request merged**